oauth2_client/models.py

```diff
@@ -1,54 +1,250 @@
+
+    @property
+    def name(self):
+        return self.app.name
+
+    @property
+    def key(self):
+        return self.app.client_id
+
+    @property
+    def secret(self):
+        return self.app.client_secret
+
+
+class TokenExpiryState:
+    VALID = 'valid'         # token not expired (including TIMEOUT_SECONDS margin)
+    UNKNOWN = 'unknown'     # no expiration info on token
+    EXPIRED = 'exired'      # token expired (including TIMEOUT_SECONDS margin)
+    ABSENT = 'absent'       # we have no token
+
 
 class AccessToken(models.Model):
     """
     Access token for the service provider.
     """
-    application = models.ForeignKey(Application, on_delete=models.CASCADE)
-    # store in JSON because OAuth2Session and BackendApplicationClient require it in this format
-    token = JSONField()
-    created_at = models.DateTimeField(auto_now_add=timezone.now, db_index=True)
+
+    """
+    Used to prevent the token expiration when the request is processed on
+    the server if the token is expiring in seconds or less at the client
+    request time.
+    """
+    TIMEOUT_SECONDS = float(60)
+
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+        related_name="%(app_label)s_%(class)s"
+    )
+    token = models.CharField(
+        max_length=255,
+        unique=True
+    )
+    raw_token = JSONField(
+        default=dict,
+        blank=True
+    )
+    token_type = models.TextField()
+    application = models.ForeignKey(
+        Application,
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+    )
+    expires = models.DateTimeField(
+        blank=True,
+        null=True
+    )
+
+    # Scope is a series of space delimited strings, eg. "read write"
+    scope = models.TextField(blank=True)
+
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    @staticmethod
+    def get_expiry_state(token):
+        """
+        Returns token expiration state.
+
+        :param token: AccessToken model object. Can be null
+        :return: string constant from TokenExpiryState
+        """
+        if token and token.is_valid_with_margin():
+            return TokenExpiryState.VALID
+        elif token and not token.expires:
+            return TokenExpiryState.UNKNOWN
+        elif token:
+            return TokenExpiryState.EXPIRED
+        elif not token:
+            return TokenExpiryState.ABSENT
+        else:
+            raise TokenError("Did we miss something? Token in crazy state.")
+
+    @staticmethod
+    def get_current_token_or_none(application_name):
+        """
+        Get the newest token for the given application or none.
+
+        :param application_name:
+        :return: AccessToken object or none
+        """
+        return (
+            AccessToken.objects
+            .filter(application__name=application_name)
```

Comment:
can it cause an issue if `application` is null?

Reply:
nope, it will just return None then. BTW that's only application_name:str, not a whole Application object